Core/include/Acts/Seeding/CandidatesForMiddleSp.ipp

```diff
@@ -0,0 +1,277 @@
+  // it may cause std::bad_alloc if we don't protect
+  if (n_high == std::numeric_limits<std::size_t>::max() or
+      n_low == std::numeric_limits<std::size_t>::max()) {
+    return;
+  }
+
+  // Reserve enough memory for all collections
+  m_storage.reserve(n_low + n_high);
+  m_storage_high.reserve(n_high);
+  m_storage_low.reserve(n_low);
+}
+
+template <typename external_space_point_t>
+inline void CandidatesForMiddleSp<external_space_point_t>::pop(
+    std::vector<std::size_t>& storage, std::size_t& current_size) {
+  // Remove the candidate with the lowest weight in the collection
+  // By construction, this element is always the first element in the
+  // collection.
+  // The removal works this way: the first element of the collection
+  // is overwritten with the last element of the collection.
+  // The number of stored elements (current_size) is lowered by 1
+  // The new first element is moved down in the tree to its correct position
+  std::swap(storage[0], storage[current_size - 1]);
+  bubbledw(storage, 0, --current_size);
+}
+
+template <typename external_space_point_t>
+inline bool CandidatesForMiddleSp<external_space_point_t>::exists(
+    const std::size_t& n, const std::size_t& max_size) const {
+  // If the element exists, it's index is lower than the current number
+  // of stored elements
+  return n < max_size;
+}
+
+template <typename external_space_point_t>
+inline float CandidatesForMiddleSp<external_space_point_t>::weight(
+    const std::vector<std::size_t>& storage, std::size_t n) const {
+  // Get the weight of the n-th element
+  return m_storage[storage[n]].weight;
+}
+
+template <typename external_space_point_t>
+inline void CandidatesForMiddleSp<external_space_point_t>::clear() {
+  // do not clear max size, this is set only once
+  // reset to 0 the number of stored elements
+  m_n_high = 0;
+  m_n_low = 0;
+  // Reset all values to default
+  m_storage.clear();
+  m_storage_high.clear();
+  m_storage_low.clear();
+}
+
+template <typename external_space_point_t>
+void CandidatesForMiddleSp<external_space_point_t>::push(
+    external_space_point_t& SpB, external_space_point_t& SpM,
+    external_space_point_t& SpT, float weight, float zOrigin, bool isQuality) {
+  // Decide in which collection this candidate may be added to according to the
+  // isQuality boolean
+  if (isQuality) {
+    return push(m_storage_high, m_n_high, m_max_size_high, SpB, SpM, SpT,
+                weight, zOrigin, isQuality);
+  }
+  return push(m_storage_low, m_n_low, m_max_size_low, SpB, SpM, SpT, weight,
+              zOrigin, isQuality);
+}
+
+template <typename external_space_point_t>
+void CandidatesForMiddleSp<external_space_point_t>::push(
+    std::vector<std::size_t>& storage, std::size_t& n, const std::size_t& n_max,
+    external_space_point_t& SpB, external_space_point_t& SpM,
+    external_space_point_t& SpT, float weight, float zOrigin, bool isQuality) {
+  // If we do not want to store candidates, returns
```

Comment:
Any feeling which of the paths is more likely? May the other of these checks be relevant for performance, again in c++20 we'd mark some of these branches with `[[likely]]`, for now maybe ordering them is useful (so in future we can easily find the likely path and mark it. Just a suggestion.

Reply:
Hard to say since it depends on the max number of seeds the user wants to keep (default is 5 as of now) and how crowded the event is.